### DISCUSSION TITLE
(NextJS) What does this mean: "Because the Redux store is shared across requests"

### DISCUSSION THREAD

**Discussion by emreakdas** [reduxjs/redux-toolkit#4314] [General]
"No global stores - Because the Redux store is shared across requests, it should not be defined as a global variable. Instead, the store should be created per request."

I came across this while reading the document, can you give an example for this scenario to occur? I did not fully understand what is meant by saying that the global store data may be contaminated.

and what problems can sharing your wishes cause?

**Comment by phryneas** [comment]
User A visits your page. Your page is executed in SSR, a global store variable is created.
User B visits your page. Your page is executed in SSR. The same global store variable is used.

Relevant knowledge: In Next.js, every client components will automatically SSR for the first render of a user session.

**Reply by emreakdas** [reply]
How do you use the same global store?

**Reply by emreakdas** [reply]
This is said for RSC, and it sounds absurd why you say why the same global store should be used in two different users.

**Reply by emreakdas** [reply]
it turns out that you say: two separate devices use the same store

**Reply by EskiMojo14** [reply]
remember, it's the same server handling both requests

**Reply by emreakdas** [reply]
the logic you say goes like this;

If I create a global variable in the nextjs project, let counter = 0;

If user a increments, user b sees 1 on input.

@EskiMojo14 did I misunderstand?

**Reply by EskiMojo14** [reply]
if that variable is on the server side, that's my understanding yes

**Reply by phryneas** [reply]
It is one long-running server process that's shared between all your users: *any* global variable (in SSR, or in RSC) will exist for the whole time that the server runs without the process being killed/restarted. A single process can handle hundreds of thousands of users requests without restarting. 

All of these users will see the same value of that global variable (e.g., a Redux store).


> This is said for RSC, and it sounds absurd why you say why the same global store should be used in two different users.

I would have assumed that the sentence directly after the sentence you are quoting,

"RSCs should not read or write the Redux store - RSCs cannot use hooks or context. They aren't meant to be stateful. Having an RSC read or write values from a global store violates the architecture of the Next.js App Router."

would make it clear that we are talking about SSR here. If obviously doesn't make sense to have a global or non-global Redux store in RSC at all, when you are neither allowed to write nor to read it.

**Reply by emreakdas** [reply]
I understand more clearly thanks @phryneas <3

**Reply by emreakdas** [reply]
@phryneas

what I really wanted to ask you was;

> User A visits your page. Your page is executed in SSR, a global store variable is created.
> User B visits your page. Your page is executed in SSR. The same global store variable is used.

You said. what kind of problem can be caused by user b using the same store? no data reading and writing is done from rsc, after all. client operations cannot change the store. what is mentioned in the document when it is said "it may be contaminated"?

Isn't it actually a situation like the store is used as initial value on the server?

**Reply by markerikson** [reply]
@emreakdas that's the exact problem. If the same store instance gets used on the server across requests, it may end up accidentally including and sending back some of user A's data to user B. Which would be _very_ bad.

**Reply by emreakdas** [reply]
@markerikson but we don't read/write to the store on the server, am I misunderstanding the situation?

**Reply by emreakdas** [reply]
In other words, the transactions we make to the store in the client are not valid on the server side. I don't understand how the transactions we make in the client will be valid on user b.

**Reply by emreakdas** [reply]
In other words, how will the store on the server change so that the data of user a can go to user b with the operation I do in the client..

 @markerikson

**Reply by markerikson** [reply]
@emreakdas please don't keep re-pinging people.

The assumption in this whole discussion is that **yes, you _are_ writing to the Redux store on the server to pre-fetch data needed by the page**.

**Reply by emreakdas** [reply]
then if I don't write anything to the store on the server, there is no problem with using a global store, am I right?

this time in your document `RSCs should not read or write the Redux store` is written this way. `No global stores` is not contrary to all these things we talked about?

**Reply by markerikson** [reply]
@emreakdas I'm not sure what you're asking at this point, tbh.

Next still generally requires the careful store creation setup because of its multi-page / SSR+SPA architecture, even if you're only using Redux on the client side, because there is no one single central place that the client app code gets initialized.

The _specific_ section you originally asked about that says "shared across requests" is talking about _if you are using a Redux store on the server to pre-fetch data_.

**Reply by emreakdas** [reply]
@markerikson Isn't the situation where requests are mixed in user a or b a situation if you are also using the store on the server side?

I am currently using it only on the client side. I do not read / write data to the store on the server side. For this reason, I do not need the store created per request, am I right?

I wanted to say this.

**Reply by markerikson** [reply]
@emreakdas That's what I'm trying to _tell_ you, but you're not listening to me :(

I specifically answered that in the previous comment:

> Next still generally requires the careful store creation setup because of its multi-page / SSR+SPA architecture, even if you're only using Redux on the client side, because there is no one single central place that the client app code gets initialized.
> 
> The specific section you originally asked about that says "shared across requests" is talking about if you are using a Redux store on the server to pre-fetch data.

**Reply by emreakdas** [reply]
@markerikson I can say that it completely sat in my head. sorry if I gave you a headache :)(

from the whole correspondence I was able to make the following inference; if we are not writing data to the redux store on the server, there is no point in creating a store per request.

**Reply by markerikson** [reply]
That is literally what I've been telling you, yes :)

But also as I've been telling you, you _do_ need to follow the setup guidelines in that page anyway _even if you're only using Redux on the client_, because of how the multi-page SPA setup works.

So, please. Just _do what that page says and you won't have to worry about this_.

**Reply by emreakdas** [reply]
got it thanks a lot :)

**Reply by phryneas** [reply]
> from the whole correspondence I was able to make the following inference; if we are not writing data to the redux store on the server, there is no point in creating a store per request.

Yes and no. 

As I said in my first comment: Next.js will execute your Client Components on the server, during SSR. So your Redux store *will* be read from during SSR, because your Client Components are executed.
Now if that's a global store shared by a million users, you only need to change anything once on accident and all future SSR renders will read from that accidentally modified store.

It's just too risky, *not* to scope a Redux store to the tree. Do it.

That also makes sure that you never use Redux in RSC: since it's not globally available, you can't even access it.

It's all about patterns that protect you from shooting yourself in the foot. It's extremely easy to have a component that's rendered from a Client Tree today and from a RSC tree tomorrow. Things accidentally move from "local" to "server". Use patterns to make that impossible from the start.

**Reply by emreakdas** [reply]
@phryneas all these operations are actually done to read and write data from the server, isn't it done to solve the problem in advance?

**Reply by phryneas** [reply]
I'm sorry I don't understand the question.

**Reply by emreakdas** [reply]
@phryneas

 Aren't all the precautions you mentioned actually so that data is not accidentally written / read to the store on the server side? When data is not read and written on the server side, doesn't the store on the server only serve as an initial value used for ssr only in client usage?
  
 If I'm only using redux on the client side. What kind of problem could that cause? Isn't the precaution just to protect the developer

**Reply by phryneas** [reply]
> If I'm only using redux on the client side. What kind of problem could that cause? Isn't the precaution just to protect the developer

No, you're not, even if you only use it in "Client Components", you *also* use it Server Side. Client Components are rendered on the server. 

Unless you have something like `if (typeof window == "undefined") return null` in your root component, your application *will* render on the server.  (And if you did this, you'd get a hydration error every page load).
It's unavoidable, a core part of Next.js.

You shouldn't use any globals in a SSR-oriented framework like Next.js. It's just proper code hygiene.

**Reply by emreakdas** [reply]
when I say I'm using it, I mean this;

I just meant I write to the store in the client and read it.

--

After all, the data on the client will not go to the server, the store on the server will not change. (In the scenario where we do not write or read a value to the store on the server).

that's what I meant.

**Reply by emreakdas** [reply]
What I mean is that if data is written to the store on the server, we do it to prevent confusion. but in my case, it will not be written to the store on the server and will not be read, I hope I can explain.

**Reply by markerikson** [reply]
I'm sorry, but this conversation is going in circles and isn't helpful at this point.  I'm going to lock the thread.
